ethstats/ethstats.go

```diff
@@ -211,93 +259,229 @@ func (s *Service) loop() {
+	Network  string `json:"net"`
+	Protocol string `json:"protocol"`
+	API      string `json:"api"`
+	Os       string `json:"os"`
+	OsVer    string `json:"os_v"`
+	Client   string `json:"client"`
+	History  bool   `json:"canUpdateHistory"`
+}
+
+// authMsg is the authentication infos needed to login to a monitoring server.
+type authMsg struct {
+	ID      string         `json:"id"`
+	Address common.Address `json:"address"`
+	Info    nodeInfo       `json:"info"`
+}
+
+// login tries to authorize the client at the remote server.
+func (s *Service) login(conn *websocket.Conn, sendCh chan *StatsPayload) error {
+	// Construct and send the login authentication
+	infos := s.server.NodeInfo()
+
+	var (
+		etherBase common.Address
+		network   string
+		protocol  string
+		err       error
+	)
+	p := s.engine.Protocol()
+	if info := infos.Protocols[p.Name]; info != nil {
+		ethInfo, ok := info.(*eth.NodeInfo)
+		if !ok {
+			return errors.New("Could not resolve NodeInfo")
+		}
+		network = fmt.Sprintf("%d", ethInfo.Network)
+		protocol = fmt.Sprintf("%s/%d", p.Name, p.Versions[0])
+	} else {
+		lesProtocol, ok := infos.Protocols["les"]
+		if !ok {
+			return errors.New("No less protocol found")
+		}
+		lesInfo, ok := lesProtocol.(*les.NodeInfo)
+		if !ok {
+			return errors.New("Could not resolve NodeInfo")
+		}
+		network = fmt.Sprintf("%d", lesInfo.Network)
+		protocol = fmt.Sprintf("les/%d", les.ClientProtocolVersions[0])
+	}
+	if s.eth != nil {
+		etherBase, err = s.eth.Etherbase()
 		if err != nil {
-			log.Warn("Stats server unreachable", "err", err)
-			time.Sleep(10 * time.Second)
-			continue
+			return err
 		}
-		// Authenticate the client with the server
-		if err = s.login(conn); err != nil {
-			log.Warn("Stats login failed", "err", err)
-			conn.Close()
-			time.Sleep(10 * time.Second)
-			continue
+	}
+	auth := &authMsg{
+		ID:      s.node,
+		Address: etherBase,
+		Info: nodeInfo{
+			Name:     s.node,
+			Node:     infos.Name,
+			Port:     infos.Ports.Listener,
+			Network:  network,
+			Protocol: protocol,
+			API:      "No",
+			Os:       runtime.GOOS,
+			OsVer:    runtime.GOARCH,
+			Client:   "0.1.1",
+			History:  true,
+		},
+	}
+	if err := s.sendStats(conn, actionHello, auth); err != nil {
+		return err
+	}
+
+	// Proxy needs a delegate send here to get ACK
+	if s.backend.IsProxy() {
+		select {
+		case signedMessage := <-sendCh:
+			err := s.handleDelegateSend(conn, signedMessage)
+			if err != nil {
+				return err
+			}
+		case <-time.After(delegateSendTimeout * time.Second):
+			// Login timeout, abort
+			return errors.New("login timed out")
 		}
-		go s.readLoop(conn)
+	}
 
-		// Send the initial stats so our node looks decent from the get go
-		if err = s.report(conn); err != nil {
-			log.Warn("Initial stats report failed", "err", err)
-			conn.Close()
-			continue
+	// Retrieve the remote ack or connection termination
+	var ack map[string][]string
+	if err := websocket.JSON.Receive(conn, &ack); err != nil {
```

Comment:
I think there's an issue here where you're doing things if there was an error, instead of doing things if there wasn't an error

maybe instead do
```
if err := websocket.JSON.Receive(conn, &ack); err != nil {
  return err
}
// everything else
```

Reply:
checked this out and it's correct. The stuff in the branch are other error conditions.